engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java

```diff
@@ -1013,6 +1013,132 @@ public void shouldSetLoopCounterVariable() {
   }
 
+  @Test
+  public void shouldCompleteBodyWhenCompleteConditionAccessNumberOfInstancesEvaluateTrue() {
+    // given
+    int completedJobs = 2;
+    String completionCondition = "= numberOfInstances = 2";
+    if ("parallel".equals(loopCharacteristics)) {
+      completionCondition = "= numberOfInstances = 3";
+      completedJobs = 1;
+    }
+    final String finalCompletionCondition = completionCondition;
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            process(miBuilder.andThen(m -> m.completionCondition(finalCompletionCondition))))
+        .deploy();
+
+    // when
+    final long processInstanceKey =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariable(INPUT_COLLECTION_EXPRESSION, INPUT_COLLECTION)
+            .create();
+
+    completeJobs(processInstanceKey, completedJobs);
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .withElementId(ELEMENT_ID))
+        .extracting(r -> tuple(r.getValue().getBpmnElementType(), r.getIntent()))
+        .containsSubsequence(
+            tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.COMPLETE_ELEMENT),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.ELEMENT_COMPLETING),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.ELEMENT_COMPLETED));
+
+    if ("parallel".equals(loopCharacteristics)) {
+      // after 1 has completed, the others must be terminated
+      assertThat(
+              RecordingExporter.records()
+                  .limitToProcessInstance(processInstanceKey)
+                  .processInstanceRecords()
+                  .withIntent(ProcessInstanceIntent.ELEMENT_TERMINATED)
+                  .withElementType(BpmnElementType.SERVICE_TASK)
+                  .count())
+          .describedAs("all non-completed service tasks have terminated")
+          .isEqualTo(2);
+    } else {
+      assertThat(
+              RecordingExporter.records()
+                  .limitToProcessInstance(processInstanceKey)
+                  .processInstanceRecords()
+                  .withIntent(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                  .withElementType(BpmnElementType.SERVICE_TASK)
+                  .count())
+          .describedAs("only 2 out of 3 sequential service tasks has activated")
+          .isEqualTo(2);
+    }
+  }
+
+  @Test
+  public void shouldCompleteBodyWhenCompleteConditionAccessNumberOfActiveInstancesEvaluateTrue() {
+    // given
+    int completedJobs = 1;
+    String completionCondition = "= numberOfActiveInstances = 0";
+    if ("parallel".equals(loopCharacteristics)) {
+      completedJobs = 3;
```

Comment:
🔧 I believe that this test case for the parallel loopCharacteristics actually works best when you complete 2 of the 3 jobs. Once the 2nd instance completes, the completionCondition evaluates to true and then terminates the last instance. The reason that the test also works with 3 is that the engine seems to run faster than the 3rd complete job command can be sent and processed.